Adding jigsaw examples

ladoramkershner opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Adding examples to the jigsaw documentation.

## Description
<!--- Describe your changes in detail -->
I am adding three examples; one bundle for a relative network, one for a grounded network (or ready for cube update), and a grounded network (ready for updating cubes).

## Related Issue
<!--- This project only accepts pull requests related to open issues (GitIssues) -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
Issue #4718 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Previously there was a single jigsaw example, which ran mostly defaults. These examples, alter the default parameters while explaining the reason for the alterations. 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
The docs were built locally and previewed before this PR was opened. 

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- - [ ] My code follows the code style of this project. -->
- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

ladoramkershner committed:
jigsaw ex KAGTC AP15; initil commit

ladoramkershner committed:
splitting long example into three

ladoramkershner commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+    </example>
+
+    <example>
+      <brief>Run of jigsaw parameterized for Kaguya Terrain Camera (TC) images and <b>an intermediate ground control</b> network covering the Apollo 15 landing site. </brief>
```
Not terribly happy with the terminology of 'intermediate ground control network' but I was not sure what else to call it. If you have a suggestion, please let me know!

jessemapel replied:
I think `intermediate` is okay. Another option could be `initial`. The more important part is you explain why it's not ready for final use yet; which you do.

ladoramkershner commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+    </example>
+
+    <example>
+      <brief>Run of jigsaw parameterized for Kaguya Terrain Camera (TC) images and <b>a final ground control</b> network covering Apollo 15 landing site. </brief>
```
Also not happy with this terminology, again suggestions are welcome.

ladoramkershner commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+      </terminalInterface>
+    </example>
+
+
```
I realize I have added no supplementary files or GUI images. Please let me know if you think adding those would enhance the examples.

jessemapel replied:
I don't think we need GUI images, but it would be nice if you could provide the bundleout.txt for each solution as it contains the human-readable information that would help contextualize the data.

ladoramkershner replied:
Can I include the entire bundleout.txt file? Each one is about ~25M, so that would be almost a Gig for all three examples. 

Alternatively I could shorten the files, just keep the full headers and a few example lines from each section?

ladoramkershner committed:
Updating Changelog

jessemapel commented on CHANGELOG.md:
```diff
@@ -52,6 +52,7 @@ Keywords when running CAMSTATS.  [#3605](https://github.com/USGS-Astrogeology/IS
 - Added additional translation files for TGO CaSSiS in order to support PSA compliant labels. [#4567](https://github.com/USGS-Astrogeology/ISIS3/issues/4567)
 - Added support for KaguyaTC SP Support data ingest. [#4668](https://github.com/USGS-Astrogeology/ISIS3/issues/4668)
 - Added examples to the jigsaw documentation. [#4718](https://github.com/USGS-Astrogeology/ISIS3/issues/4718)
+- Added three examples to the jigsaw documentation covering relative and ground control bundles. [#4718](https://github.com/USGS-Astrogeology/ISIS3/issues/4718)
```
I think we can just go with Stuart's original change log entry for all of these example updates. It's odd to have two changelog entries pointing to the same issue.

ladoramkershner replied:
I removed my entry, Stuart verbiage encompasses the addition of these examples

jessemapel commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+        <p>
+         A relative network is a network that has no points connected to a ground source, but connects overlapping images together. Since there is no connection to 
+         ground in the network, only camera specific parameters can be solved in the bundle. The required information to turn on spacecraft, radius, or other such
+         bundle solve parameters is not recommended.
```
I'm not sure if this paragraph is entirely correct. You can certainly still solve for all of these things, as long as you have enough observations for the number of unknowns, but they will all be determined relative to the spacecraft. So, the ground solution relative to the spacecraft will be correct, but the correctness of the ground solution relative to the rest of the body is unknown. You can come back and use tools like PC align to correct for this type of error.

There is likely a more technical set of terms for this, I'll reach out to Brent.

ladoramkershner replied:
I guess I am under the impression that within ISIS, if you do not have ground in a network then solving for spacecraft position is not really worth the effort. But I would really appreciate more certain terminology, let me know if you hear from Brent

ladoramkershner replied:
For now I will not NOT recommend not solving for other variables, just state that we are solving for camera parameters.

jessemapel commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+        </p> 
+        <p>
+         This relative bundle turns on and parameterizes the camera twist and camera acceleration solve parameters. Turning on camera twist allows for the bundle to solve 
+         for the cameras rotation around the bore sight axis and does not require parameterization. Setting the camera solve parameter to acceleration is useful for 
```
`and does not require parameterization` what does this mean? Is this saying that solving for twist is optional?

ladoramkershner replied:
I mean twist is a flag and does not have a corresponding uncertainty estimation requirement. I will update to be more clear.

jessemapel commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+         This relative bundle turns on and parameterizes the camera twist and camera acceleration solve parameters. Turning on camera twist allows for the bundle to solve 
+         for the cameras rotation around the bore sight axis and does not require parameterization. Setting the camera solve parameter to acceleration is useful for 
+         line scan imagers and requires uncertainties to be set for the angles (deg), angular velocity (deg/s), and angular accelerations (deg/s**2). These values were 
+         set with increasing constraint because of the increasing affect alterations of higher derivative parameters have on the bundle solution.
```
Change `higher derivative parameters` to `higher order parameters`. We're still just computing first partials with respect to all of these parameters, the acceleration is a higher order term in the pointing polynomials.

jessemapel commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+        <p>
+         The ‘overexisting’ flag allows the bundle solution to approximate the camera solution with a polynomial function as opposed to storing every individual measurement. 
+         This saves time and memory required for the bundle solution at the expense of accuracy. If one required the accuracy, overexisting could be used until the bundle 
+         is converging in a satisfactory way and then turned off for the final solution.
```
This paragraph is incorrect. The bundle solves for a polynomial regardless of `overexisting` or not. When you do not have `overexisting` on, the bundle starts by fitting a polynomial to the a-priori ephemerides and then just uses that polynomial to compute all ephemerides. If you enable `overexisting`, it instead saves the a-priori ephemerides and begins with a zero polynomial. The a-priori ephemerides are then added to ephemerides computed from the polynomial that the bundle solves for. Here's a simple example:

### overexisting off

Initial data : {(1,1), (2, 4), (3, 9), (4, 16)}

fitting a linear polynomial we get f(x) = 5x - 5

then, we throw the initial data away and we now compute all ephemerides using our function, f. So, {(1, f(1)), (2, f(2), ...}

The bundle will compute corrections to the coefficient of f.

### overexisting on

Initial data : {(1,1), (2, 4), (3, 9), (4, 16)}

We start with a zero polynomial, f(x) = 0x + 0

Now, we keep the original data and add the result of f to it. So, {(1, 1+f(1)), (2, 4+f(2)), (3, 9+f(3)), ...}.

The bundle still computes corrections to our function f and because the initial data is just a constant offset, it doesn't impact our partial computations in the bundle.

ladoramkershner replied:
Oh yeah, I was very wrong. What is the benefit of using overexisting then?

jessemapel commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+      </description>
+      <terminalInterface>
+        <commandLine>
+        jigsaw fromlist=cubes.lis cnet=relative_reg_jig4_edit.net onet=relative_reg_jig5.net update=no file_prefix=jig5 sigma0=1.0e-10 maxits=10 \
```
Add a not above that you have `update=no` because you are still evaluating the solution and do not want to apply it to your images just yet.

jessemapel commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+        </p>
+        <p>
+         Ground points are weighted more heavily in the bundle than relative control points and adding parameters adds more complexity to the bundle solution. Therefore, it 
+         is common to create and smith a network with only relative points and add ground point in AFTER the relative network (and its bundle solution) is of sufficient quality. 
```
I would avoid the word `smith` as it could be confused with `smithed` kernels that are produced from bundle adjusted images. I would instead use something like `refined` or `improve`.

jessemapel commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+        </p>
+        <p>
+         Ground points are weighted more heavily in the bundle than relative control points and adding parameters adds more complexity to the bundle solution. Therefore, it 
+         is common to create and smith a network with only relative points and add ground point in AFTER the relative network (and its bundle solution) is of sufficient quality. 
```
I would bold `after` instead of using all caps. In the brief description bold is used for emphasis but here all caps are.

ladoramkershner replied:
I switched because I was having trouble with bold not showing up, I will switch it to be consistent (which is a good suggestion) and hopefully I can figure out why the bold is not showing when I build the docs locally.

jessemapel commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+        </p>
+        <p>
+         In addition to the previous solve parameters, this bundle turns on the point radius and spacecraft position solve parameters. Radius and spacecraft position 
+         are two parameters which can only be solved for when ground points are present in the bundle solution since they are concerned with absolute positioning.
```
See my comment on the previous example about relative solutions vs grounded solutions. You can 100% still solve for these things and they will be relatively correct without ground.

ladoramkershner replied:
I will just delete the sentence "Radius and spacecraft position are two parameters which can only be solved for when ground points are present in the bundle solution since they are concerned with absolute positioning"

jessemapel commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+         Applying the point radius solve parameter requires the point_radius_sigma to be set. This value is a representation the uncertainty (in meters) of the 
+         cameras pointing corresponding to the correct elevation on the shape model. Often this value can be set, and the appropriateness of the set value can be 
+         checked using the 'POINTS DETAIL' section of the bundleout.txt file output by jigsaw. If the radius total correction of points exceeds the provided sigma, 
+         the uncertainty may need to be increased.
```
Be careful correlating these too closely. We've had users confused about the sigmas not being hard constraints. They are a-priori uncertainties, or standard deviations, not hard constraints. So, it's expected that some amount of points have corrections higher than their sigmas, in fact in a perfect world, we'd expect about a third of the points to have corrections higher than 1 sigmas.

jessemapel commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+         Applying the space craft position solve parameter requires an uncertainty estimation through spacecraft_position_sigma, while [file_prefix]_bundleout_images.csv 
+         holds the resulting image adjustments in the X Correction, Y Correction, and Z Correction columns. The overhermite flag allows a polynomial to estimate the 
+         spacecraft position, like the overexisting flag estimates the camera pointing with a polynomial. As such, this option saves memory and time at the expense of 
+         some loss in accuracy and this loss can be mitigated by switching it off for the final run.
```
See my comments in the previous example. using `overexisting` or `overhermite` actually takes more memory than just solving for a polynomial as we have to store the a-priori ephemerides and the polynomial.

jessemapel commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+         your update cubes camera pointing (or any kernels resulting from these updated camera pointings).
+        </p>
+        <p>
+         If you want to double check the update was completed, see cathist or catlab (search for ‘Jigged’).
```
It would be great to provide an example of a table that has been updated and one that hasn't. Like a before and after.

jessemapel commented on isis/src/control/apps/jigsaw/jigsaw.xml:
```diff
@@ -1513,5 +1513,106 @@
+         This last example is of a final jigsaw run of a grounded network. In this example all network adjustments are done, the bundle is converging, the resulting 
+         residuals are acceptable, and therefore we are ready to update the camera pointing on the cubes. In the final run you can turn off overexisting and overhermite, 
+         if you want to regain the accuracy, but note that the bundle solution will differ from your previous run. Therefore, it is advisable to conduct one intermediate 
+         run to check the bundle is acceptable without these flags before updating the cubes. Finally, during the final run we turn on the error propagation flag. This 
```
I would recommend not changing any parameters besides `update` in the final solution. You should check that you like the whole solution before you save it. It's kind of a hassle, but there's an issue open already to allow us to apply a previous solution independently. #4474

ladoramkershner committed:
removing CHANGELOG entry

ladoramkershner committed:
addressing comments round1

ladoramkershner commented:
Currently I have addressed all the comments I 👍 'd (or attempted to address them). I would still like to add 
- a table before and after cube update
- the bundleout.txt (just wanting to know how big those files should realistically be)

ladoramkershner committed:
adding bundlout.txt examples to documents

ladoramkershner committed:
adding table summaries before and after update

jessemapel committed:
Small language tweaks for overhermite/existing

jessemapel reviewed (approved):
I made a few minor tweaks to the language, but this is ready to go.
